enhancements/installer/gcp-ipi-shared-vpc.md

```diff
@@ -0,0 +1,260 @@
+automation of an install to a shared VPC.
+- As a user, I want the installer to use existing resources in the Host Project
+  so that I can create infrastructre in advance and minimize the privileges to
+the Host Project given to the installer when installing to a shared VPC.
+
+### Goals
+
+- Enable users to deploy GCP clusters with Shared VPC networking using the IPI
+  workflow.
+- Enable users to provide service account(s) with sufficient permissions in the
+  Host Project to enable provisioning the majority of the cluster resources.
+- Enable users to alternatively provide pre-provisioned resources to enable
+  minimal permission requirements in the Host Project.
+
+### Non-Goals
+
+- Provisioning and/or configuring the VPC, Network, and Subnetworks are to be
+  done manually by the user prior to running the installer.
+- Provisioning and/or configuring the Service Accounts for the Installer and
+  Cluster are to be done manually by the user prior to running the installer.
+
+## Proposal
+
+The installer should be able to install a GCP IPI cluster using Shared VPC
+networking. This requires some resources to be created and configured in
+advance, such as the network and subnets. Other resources can be created in
+advance when the necessary host project permissions are not desired, such as
+the firewall rules and/or DNS configuration. The installer, and some operators,
+will be updated to enable this functionality. They will become capable of
+creating and managing resources that exist in a different (Host) project from
+the cluster (Service Project). They will also become capable of consuming
+pre-existing firewall and/or DNS configurations.
+
+### Workflow Description
+
+Prior to executing the installer, the Shared VPC, Network, and Subnets will be
+manually created in the Host Project and configured for Shared VPC (XPN)
+configuration with the Service Project. A service account will also be created
+in the Service Project and granted sufficient permissions in the Host Project.
+When minimal permissions are desired in the Host Project, additional resources
+will be manually created and configured in the Host Project to include:
+firewall rules, api compute address, ingress compute address, and the api,
+api-int, and \*.apps DNS records.
+
+After all of these resources are created, they will be used in a manually
+created install-config.yaml to configure the cluster accordingly. At a minimum,
+the credentialsMode will be set to Passthrough, and the network,
+computeNetwork, controlPlaneNetwork, and networkProjectID will be configured to
+reference the Shared VPC network. When minimial permissions are desired in the
+Host Project, the additional parameters will be configured to enable the
+cluster to consume them as well.
+
+The installation process can then proceed as any other IPI workflow by running
+the create manifests, create ignition, and/or create cluster sub-commands.
+
+### API Extensions
```

Comment:
Are there any changes to openshift APIs that reside in cluster? This all seems to be install config changes if I'm understanding correctly.

Have we got some description of the ingress API changes required for this?

Reply:
There are API changes, as mentioned in this document... as well as changes for operators. I was hoping to have the team responsible for that work help iron those details out.

Reply:
For reference, the API changes are the sections above: api.dns, api.ingress

Reply:
Cool, once you've worked out what new API fields will be required and what you'd like those to look like (seems maybe only 1 or 2?), can you write up a small Go snippet for them to represent the actual code changes to the API code including documentation and kubebuilder tags based on the [API conventions docs](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md)

Reply:
I've removed the need for any API code changes.